src/compiler/scala/tools/nsc/Global.scala

```diff
@@ -401,7 +405,7 @@ class Global(var currentSettings: Settings, reporter0: Reporter)
 
     final def withCurrentUnit(unit: CompilationUnit)(task: => Unit): Unit = {
```

Comment:
if this is the entry point for the `task`, maybe the CU can hold the ThreadLocal storage

This would allow for cleanup at the end of the run, and a common scheduling /reporting/recovery

Its doesn't seem to be re-entrant. Not sure about compileLate though

maybe something like
```
    def run() {
      assertOnMainThread()
      val pending = new ListBuffer[(CompilationUnit, Future[Unit])]

      @tailrec def runParallel(unit: CompilationUnit): Unit = {
        try {
          val task = Future(applyPhase(unit))
          pending += ((unit, task))
        } catch {
          case CantScheduleAsFull =>
            endPhase(pending.remove(0))
            runParallel(unit)
        }
      }
      def startPhase(unit: CompilationUnit) = {
        if (parallel) runParallel(unit)
        else endPhase(unit, Future.fromTry(scala.util.Try(apply(unit))))
      }
      def endPhase(unit: CompilationUnit, complete: Future[Unit]) = {
        Await.result(complete, Duration.Inf)
        process buffered data from CU
      }
      
      echoPhaseSummary(this)
      currentRun.units foreach startPhase
      pending foreach {
        case ((unit, complete) => endPhase(unit, complete)
      }
    }
    def parallel = false
```